turbopack/crates/turbo-tasks-fs/src/lib.rs

```diff
@@ -72,6 +72,81 @@ use crate::{
+///
+/// The behaviour of the file system changes depending on the OS, and indeed sometimes the FS
+/// implementation of the OS itself.
+///
+/// - On windows the limit for normal file paths is 260 characters, a holdover from the DOS days,
+///   but rust will opportunistically rewrite paths to 'UNC' paths for supported path operations
+///   which can be up to 32767 characters long.
+/// - On macOS, the limit is 255 characters traditionally, but after personal testing it seems to be
+///   around 298 on my (@arlyon) machine.
+/// - On linux, the limit differs between kernel (and by extension, distro) and filesystem. Individual files can be up to 255 bytes, while
+///   paths can be up to 4096 _on EXT4_. However, despite advertising these in `limits.h`, some filesystems actually support larger lengths.
+///   Fedora for example explicitly says not to program against the limit on linux (https://docs.fedoraproject.org/en-US/defensive-coding/tasks/Tasks-File_System/#sect-Defensive_Coding-Tasks-File_System-Limits)
+///
+/// Realistically, the output path lengths will be the same across all platforms, so we need to set
+/// a conservative limit and be particular about when we decide to bump it. Here we have opted for
+/// 255 characters, because it is the shortest of the three options.
+fn validate_path_length<'a>(path: &'a Path) -> Result<Cow<'a, Path>> {
+    /// Here we check if the path is too long for windows, and if so, attempt to canonicalize it
+    /// to a UNC path.
+    #[cfg(target_family = "windows")]
+    fn validate_path_length_inner<'a>(path: &'a Path) -> Result<Cow<'a, Path>> {
+        const MAX_PATH_LENGTH_WINDOWS: usize = 260;
+        const UNC_PREFIX: &str = "\\\\?\\";
+
+        if path.as_os_str().starts_with(UNC_PREFIX) {
+            return path.into();
+        }
+
+        if path.as_os_str().len() > MAX_PATH_LENGTH_WINDOWS {
+            let new_path = normalize_path(&unix_to_sys(&path.to_string_lossy()))
+                .ok_or_else(|| anyhow!("file is too long, and could not be normalized"))?;
+            return new_path.into();
+        }
+
+        Ok(path.into())
+    }
+
+    /// Here we are only going to check if the total length exceeds, or the last segment exceeds.
+    /// This heuristic is primarily to avoid long file names, and it makes the operation much
+    /// cheaper.
+    #[cfg(not(target_family = "windows"))]
+    fn validate_path_length_inner<'a>(path: &'a Path) -> Result<Cow<'a, Path>> {
+        // on most linux distros this is likely a lot larger than 1016, but macOS is *special*
+        const MAX_PATH_LENGTH: usize = 1016;
+
+        // check the last segment (file name)
+        if path
+            .file_name()
+            .map(|n| n.as_encoded_bytes().len())
+            .unwrap_or(0)
+            > MAX_FILE_NAME_LENGTH_UNIX
+        {
+            return Err(anyhow!(
+                "file name is too long (exceeds {} bytes)",
+                MAX_FILE_NAME_LENGTH_UNIX
+            ));
```

Comment:
Anyhow has a shorthand macro for this

```suggestion
            anyhow::bail!(
                "file name is too long (exceeds {} bytes)",
                MAX_FILE_NAME_LENGTH_UNIX
            );
```